### PR TITLE
eliminate parallelStream

### PR DESCRIPTION
**Pull request by noblepaul**
while we were testing the RC1 for Solr 9.1 release, we observed that there is a significant slowdown in shutting down nodes. upon further investigation we realized that there were threads hanging in ForkJoinPool.commonPool. These threads were hanging forever even after 3 minutes of shutting down the process. A single threaded operation is fast enough

forkjoinpool

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -2865,39 +2865,57 @@ public boolean checkIfCoreNodeNameAlreadyExists(CoreDescriptor dcore) {
...
-              replicasToDown.add(replica.getName());
-            }
-          });
-          try {
-            PerReplicaStatesOps.downReplicas(replicasToDown,
-                PerReplicaStates.fetch(coll.getZNode(), zkClient, coll.getPerReplicaStates())).persist(coll.getZNode(), zkClient);
-          } catch (KeeperException | InterruptedException e) {
-            throw new RuntimeException(e);
+      try {
+        // Create a concurrently accessible set to avoid repeating collections
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: whitespace is a bit off -- entire method over-indented at the top level, makes review unnecessarily difficult

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 2]
```diff
@@ -2865,39 +2865,57 @@ public boolean checkIfCoreNodeNameAlreadyExists(CoreDescriptor dcore) {
    * @param nodeName to operate on
    */
   public void publishNodeAsDown(String nodeName) {
-    log.info("Publish node={} as DOWN", nodeName);
-    try {
-      // Create a concurrently accessible set to avoid repeating collections
-      Set<String> processedCollections = ConcurrentHashMap.newKeySet();
-      cc.getCoreDescriptors().parallelStream().forEach(cd -> {
-        DocCollection coll = zkStateReader.getCollection(cd.getCollectionName());
-        if (processedCollections.add(coll.getName()) && coll.isPerReplicaState()) {
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2874]: Do we no longer care whether `coll.isPerReplicaState()`?

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 3]
```diff
@@ -2865,39 +2865,57 @@ public boolean checkIfCoreNodeNameAlreadyExists(CoreDescriptor dcore) {
...
-                PerReplicaStates.fetch(coll.getZNode(), zkClient, coll.getPerReplicaStates())).persist(coll.getZNode(), zkClient);
-          } catch (KeeperException | InterruptedException e) {
-            throw new RuntimeException(e);
+      try {
+        // Create a concurrently accessible set to avoid repeating collections
+        Set<String> collectionsInThisNode = new HashSet<>();
+        for (CoreDescriptor cd : cc.getCoreDescriptors()) {
+          if (cd.getCloudDescriptor() != null
+              && cd.getCloudDescriptor().getCollectionName() != null) {
+            collectionsInThisNode.add(cd.getCloudDescriptor().getCollectionName());
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: I suspect that this fix could be done more succinctly by keeping the majority of existing code, with two changes:
1. parallelStream => stream
2. use `cd.getCloudDescriptor().getCollectionName()` to get collection name pre-deduping (as you've done here), deferring `zkStateReader.getCollection()` to ensure it's invoked only once per collection.

smaller diff in and of itself doesn't matter; but it serves to clarify what's actually changing, and reduces the likelihood of unnecessary future merge conflicts.

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 4]
```diff
@@ -2865,39 +2865,57 @@ public boolean checkIfCoreNodeNameAlreadyExists(CoreDescriptor dcore) {
    * @param nodeName to operate on
    */
   public void publishNodeAsDown(String nodeName) {
-    log.info("Publish node={} as DOWN", nodeName);
-    try {
-      // Create a concurrently accessible set to avoid repeating collections
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2870]: nit: remove "concurrently" from comment, as no longer applicable

**Review by magibney — CHANGES_REQUESTED** [review]
Thanks for catching this! I left a few questions/suggestions.

It's a little unsettling that `parallelStream()` resulted in threads actually hanging -- parallel being slower or unnecessary is one thing; hung threads is another! I wonder if it'd be worth pulling at that thread (independent of this change), since in principle it should be possible to invoke these operations in parallel, no?

EDIT: To more clearly demonstrate what I mean, I opened #201 against this branch; the diff against `release/8.8` is contained in the single commit: 2a044bbf96fa158ab4026059ed2dd90e2a2f9b3c

**Comment by magibney** [comment]
mostly LGTM. Formatting changes seem fine as they're mostly just pulled in from upstream, but it'd still be good to fix the over-indent.

To summarize, the substantial changes this introduces are: null check, swap parallelStream for serial forEach.

The only other thing that I definitely think should be adjusted (and fixed upstream too, unless I'm mistaken?) is to restore the conditional on `coll.isPerReplicaState()` -- as it stands this would create PRS entries for non-PRS collections, which even if they're ignored is probably not desirable.

**Comment by magibney** [comment]
My preference would be to merge #201 into this branch and then merge into `release/8.8`.

**Review by patsonluk — APPROVED** [review]
LGTM!

**Review by magibney — APPROVED** [review]

**Comment by magibney** [comment]
Thanks again for raising this, @noblepaul ! Went ahead with merging with the goal of getting this deployed asap.

**Pull request merged**